### ISSUE TITLE
Stable Diffusion for text2image task with DDIMscheduler, fixed weight, prompts and random seed but different inference steps would generate dissimilar images 

### ISSUE THREAD

**Issue by zhuyu-cs** [huggingface/diffusers#3481]
### Describe the bug

I fix the random seeds and prompts, only change the number of inferences, and use DDIMscheduler with eta=0. But I find that the generated images are very different with different inference steps.

### Reproduction

```
import torch
from diffusers import StableDiffusionPipeline, EulerAncestralDiscreteScheduler, DDIMScheduler, DPMSolverMultistepScheduler
import os
import numpy as np 

def set_seeds(seed=0):
    torch.manual_seed(seed)
    np.random.seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)

set_seeds(0)

model_id = 'runwayml/stable-diffusion-v1-5'
pipe = StableDiffusionPipeline.from_pretrained(model_id, torch_dtype=torch.float16)
pipe.scheduler = DDIMScheduler.from_config(pipe.scheduler.config)

pipe = pipe.to("cuda")
generator = torch.Generator(device='cuda').manual_seed(0)
prompt = "an astronaut riding a horse on mars, hd, dramatic lighting"
os.makedirs('./gpu_DDIM_results', exist_ok=True)

for steps in range(20, 75, 5):
    image = pipe(prompt=prompt, 
                height=512,
                width=512,
                num_inference_steps=steps,
                negative_prompt="",
                generator=generator,
                guidance_scale=7.5).images[0]
    filename = f'steps{str(steps)}.png'  
    image.save('./gpu_DDIM_results/'+filename)

```

### System Info

```
python               3.9.16
diffusers            0.16.0
accelerate          0.18.0
tokenizers          0.13.3
torch               1.13.1+cu117
torchaudio          0.13.1+cu117
torchvision         0.14.1+cu117
transformers              4.25.1
```
Actually, I only create a new enviroment and using following two line to generate results
```
conda install -c conda-forge diffusers
pip install accelerate
pip install transformers
```

**Comment by rupertmenneer** [comment]
That is in some ways to be expected when the difference in steps is 5, 20, and 75 such as in your code snippet. If you think about what the model is doing, it is trying to go from 'noise' to 'image' space in the number of steps provided, in 5 steps it has to jump quite dramatically through the diffusion denoising process. At 75 steps it can make much smaller steps. This will mean the output will be quite drastically different.

What happens you try producing samples with the same seed and steps are 49, 50 and 51. They should be different but quite similar?

**Comment by rupertmenneer** [comment]
@zhuyu-cs I thought I would give it a go and I can see you are right! Very different results even with one step apart using your code snippet

![image](https://github.com/huggingface/diffusers/assets/71332436/222523e5-81fd-4005-9b7b-057d7bdcba63)

However, a small addition of setting the random seed to be 0 for every time you enter your for loop produces much similar results

![image](https://github.com/huggingface/diffusers/assets/71332436/8168d322-bd21-485e-92d2-8d2fd1e87005)

e.g.
```
for steps in range(45, 50, 1):
    generator = generator.manual_seed(0)
    image = pipe(prompt=prompt, 
                height=512,
                width=512,
                num_inference_steps=steps,
                negative_prompt="",
                generator=generator,
                guidance_scale=7.5).images[0]
```

Hope that makes sense and resolves your problem :)

**Comment by zhuyu-cs** [comment]
wow, thanks for the simple suggestion, the changed code did produce the desired results. I'm not quite sure what the real reason behind this is, though. From my understanding, the `torch.Generator` just provides the function to generate random noise, so it should be the same when placed outside and inside the `for` loop. Any better explanation would be appreciated!

**Comment by zhuyu-cs** [comment]
Such a strategy seems to fail when I change the prompts, e.g, prompt :

**Comment by zhuyu-cs** [comment]
The solution is indeed correct! After identifying the most fundamental idea of this solution, this issue will be closed.

**Comment by rupertmenneer** [comment]
My understanding of this is quite simple, the `torch.Generator` object will produce deterministic randomness but is never 'reset' in your code snippet. For example it might produce the following random numbers in a list:

[8, 4, 7, 2, 3, 4, 5, 6, 2]

However, if you reset the random seeds in between for loops then you'd get behaviour such as this:

[8, 4, 7, (reset) 8, 4, 7, (reset) 8, 4, 7]

I can see why you are confused though, but imagine having a class with a variable `x = 0` and passing that to your function in your for loop, without you resetting this at the start of every loop you can't guarantee that the function you pass your class to won't change the variable and stop it being 0 (a.k.a resetting your seed) at the beginning of the loop! 

```
class Data():
    def __init__(self):
        self.x = 0

def do_something(data_class):
    data_class.x += 1

data_class = Data()
for i in range(5):
    do_something(data_class)

print(data_class.x)
```

**Comment by zhuyu-cs** [comment]
Tanks for your explanation！It is really clear for understanding.
